feat: add shadow variable search post (#102)
mlr-org/mlr3website: mlr-org/gallery/optimization/2023-01-23-shadow-variable-search-on-pima/index.qmd

---
title: "Shadow Variable Search on the Pima Indian Diabetes Data Set"
description: |
  Select features with a few lines of code.
categories:
  - feature selection
  - classification
author:
  - name: Marc Becker
    url: https://github.com/be-marc
date: 2023-01-23
bibliography: ../../bibliography.bib
knitr:
  opts_chunk:
    R.options:
      datatable.print.nrows: 6
---

{{< include ../_setup.qmd >}}

```{r 2023-01-23-shadow-variable-search-on-pima-001}
#| include: false

lgr::get_logger("mlr3")$set_threshold("warn")
lgr::get_logger("bbotk")$set_threshold("warn")
set.seed(1)
future::plan("multisession")
```

# Scope

> Feature selection is the process of finding an optimal set of features to improve the performance and robustness of machine learning algorithms.

Reviewer: ```suggestion
Feature selection is the process of finding an optimal set of features to e.g. improve the performance or robustness of machine learning algorithms.
```

> Feature selection is the process of finding an optimal set of features to improve the performance and robustness of machine learning algorithms.

Reviewer: Hier vllt. noch kurz auf die unterschiedlichen Wege eingehen mit denen in mlr3 feature selection gemacht werden kann. Also filter vs wrapper.

Reviewer: Und vor allem erklären was die generelle Idee von fselect ist, also dass man ML algos evaluiert und basierend auf den erbenissen neue features selektiert und am Ende dann normalerweise das beste erbenis nimmt.

In this article, we introduce the `r ref("mlr3fselect::fselect()")` function and the `r ref("mlr_fselectors_shadow_variable_search", "shadow variable search")` algorithm [@wu_controlling_2007] for a quick and easy feature selection.
As an example, we will search for the optimal set of features for a `r ref("mlr_learners_classif.svm", "support vector machine")` on the `r ref("mlr_tasks_pima", text = "Pima Indian Diabetes")` data set.
We assume that you are already familiar with the basic building blocks of the mlr3 ecosystem.
Some knowledge about `r ref_pkg("mlr3pipelines")` is beneficial but not necessary to understand the example.

# Task and Learner

The objective of the `r ref("mlr_tasks_pima", text = "Pima Indian Diabetes")` data set is to predict whether a person has diabetes or not.
The data set includes 768 patients with 8 measurements.

```{r 2023-01-23-shadow-variable-search-on-pima-002}
#| message: false

library(mlr3verse)

task = tsk("pima")
```

```{r 2023-01-23-shadow-variable-search-on-pima-003}
#| code-fold: true
#| column: page
#| fig-width: 14
#| fig-height: 4
#| fig.cap: "Distribution of the features in the Pima Indian Diabetes data set."
#| warning: false

> library(ggplot2)

Reviewer: ```suggestion
library(ggplot2)
library(data.table)
```

task = tsk("pima")
data = melt(as.data.table(task), id.vars = task$target_names, measure.vars = task$feature_names)

ggplot(data, aes(x = value, fill = diabetes)) +
  geom_density(alpha = 0.5) +
  facet_wrap(~ variable, ncol = 8, scales = "free") +
  scale_fill_viridis_d(end = 0.8) +
  theme_minimal() +
  theme(axis.title.x = element_blank())
```

The data set contains missing values.

```{r 2023-01-23-shadow-variable-search-on-pima-004}
task$missings()
```

Support vector machines cannot handle missing values.
We impute the missing values with the `r ref("mlr_pipeops_imputehist", "histogram imputation")` method.

```{r 2023-01-23-shadow-variable-search-on-pima-005}
learner = po("imputehist") %>>% lrn("classif.svm", predict_type = "prob")
```

# Feature Selection

The `r ref("mlr3fselect::fselect()")` function controls and executes the feature selection.

> The function internally creates an `r ref("mlr3fselect::FSelectInstanceSingleCrit", "fselect instance")` and executes the feature selection directly.

Reviewer: ```suggestion
The function internally creates an `r ref("mlr3fselect::FSelectInstanceSingleCrit")` and executes the feature selection directly.
```

> The `method` argument specifies the feature selection algorithm.

Reviewer: Wie schon mein Kommentar in der introduction: Hier muss zuerst geklärt werden was die grobe struktur von den fselector dingern ist bevor man ueber unterschiedliche Methoden redet. Hier ist noch gar nicht klar was hier abgeht finde ich.

> We use the wrapper method `r ref("mlr_fselectors_shadow_variable_search", "shadow variable search")`.

Reviewer: https://mlr-org.com/fselectors.html erwähnen

Shadow variable search adds permutated copies to the data set for each feature.

> While optimizing, it adds features until a shadow variable is selected.

Reviewer: In welcher Reigenfolge?

Reviewer: Was heißt "until a shadow variable is selected"?
Gut wäre ein konkretes beispiel

> The result is the last feature set without a shadow variable.

Reviewer: Also ich habs leider nicht so gecheckt. Ich denke das wichtigsten bei dem Post ist den algorithmus nochmal besser zu erklären.

Reviewer: Sollte es vielleicht heißen: The result are all the feature sets before a shadow variable was selected? Das result ist ja dann die beste performance von den evaluierten feature sets oder?

> The `r ref("Resampling", text = "resampling strategy")` and `r ref("Measure", text = "performance measure")` determine how the performance of a model is evaluated.

Reviewer: Inwiefern geht denn das measure bei der feature selection überhaupt ein? Das ist mir auch noch nicht klar.
